fullrt/dht.go

```diff
@@ -831,40 +831,58 @@ func (dht *FullRT) Provide(ctx context.Context, key cid.Cid, brdcst bool) (err e
 }
 
 func (dht *FullRT) execOnMany(ctx context.Context, fn func(context.Context, peer.ID) error, peers []peer.ID) int {
+	if len(peers) == 0 {
+		return 0
+	}
+
 	putctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	waitAllCh := make(chan struct{}, len(peers))
+	waitFailCh := make(chan struct{})
+	waitSuccessCh := make(chan struct{})
 	numSuccessfulToWaitFor := int(float64(len(peers)) * dht.waitFrac)
-	waitSuccessCh := make(chan struct{}, numSuccessfulToWaitFor)
 	for _, p := range peers {
 		go func(p peer.ID) {
 			fnCtx, fnCancel := context.WithTimeout(putctx, dht.timeoutPerOp)
 			defer fnCancel()
 			err := fn(fnCtx, p)
 			if err != nil {
-				logger.Debug(err)
+				if fnCtx.Err() != nil {
+					logger.Debugw("error executing operation on peer: %w", err)
+				}
+				waitFailCh <- struct{}{}
 			} else {
 				waitSuccessCh <- struct{}{}
 			}
-			waitAllCh <- struct{}{}
 		}(p)
 	}
 
-	numSuccess, numDone := 0, 0
-	t := time.NewTimer(time.Hour)
-	for numDone != len(peers) {
+	var numFail, numSuccess int
+	var t *time.Timer
+	var timeoutChan <-chan time.Time
+	var hasTimedOut bool
+
+	for numFail+numSuccess != len(peers) {
 		select {
-		case <-waitAllCh:
-			numDone++
+		case <-waitFailCh:
+			numFail++
 		case <-waitSuccessCh:
```

Comment:
To reduce complexity here we can probably just combine these into a single `chan error` and count a failure if we get an error and a success otherwise.

Reply:
Done